pydeseq2/DeseqStats.py

```diff
@@ -151,15 +172,100 @@ def summary(self):
         # Store the results in a DataFrame, in log2 scale for LFCs.
         self.results_df = pd.DataFrame(index=self.dds.dispersions.index)
         self.results_df["baseMean"] = self.base_mean
-        self.results_df["log2FoldChange"] = self.LFCs[
-            self.dds.design_matrix.columns[-1]
+        self.results_df["log2FoldChange"] = self.LFCs.iloc[
+            :, self.contrast_idx
         ] / np.log(2)
         self.results_df["lfcSE"] = self.SE / np.log(2)
         self.results_df["stat"] = self.statistics
         self.results_df["pvalue"] = self.p_values
         self.results_df["padj"] = self.padj
 
-        return self.results_df
+        print(
+            f"Log2 fold change & Wald test p-value: "
+            f"{self.contrast[0]} {self.contrast[1]} vs {self.contrast[2]}"
+        )
+        display(self.results_df)
+
+    def run_wald_test(self):
+        """Perform a Wald test.
+
+        Get gene-wise p-values for gene over/under-expression.`
+        """
+
+        n = len(self.LFCs)
+        p = self.dds.design_matrix.shape[1]
+
+        # Raise a warning if LFCs are shrunk.
+        if self.shrunk_LFCs:
+            print(
+                "Note: running Wald test on shrunk LFCs. "
+                "Some sequencing datasets show better performance with the testing "
+                "separated from the use of the LFC prior."
+            )
+
+        # Get the LFC column corresponding to the design variable.
+        if self.contrast is None:
+            self.contrast_idx = self.LFCs.columns.get_loc(
+                self.dds.design_matrix.columns[-1]
+            )
+        else:
+            # If the design matrix of dds is unexpanded, only one of the two levels
+            # will have a corresponding column
+            alternative_level = "_".join([self.contrast[0], self.contrast[1]])
+            try:
+                self.contrast_idx = self.LFCs.columns.get_loc(alternative_level)
+            except KeyError:
+                alternative_level = "_".join([self.contrast[0], self.contrast[2]])
+                self.contrast_idx = self.LFCs.columns.get_loc(alternative_level)
+                self.LFCs.iloc[:, self.contrast_idx] *= -1
+                self._change_lfc_sign = True
+                pass
+
+        # Compute means according to the LFC model.
+        mu = (
+            np.exp(self.dds.design_matrix @ self.LFCs.T)
+            .multiply(self.dds.size_factors, 0)
+            .values
+        )
+
+        # Set regularization factors.
+        if self.prior_disp_var is not None:
+            D = np.diag(1 / self.prior_disp_var**2)
```

Comment:
same as above, could you change D for more informative name?

Reply:
Thanks, done in [e536451](https://github.com/owkin/PyDESeq2/pull/32/commits/e5364518043f4edf9dccaf314ad8020bc7ba2b87)